lib/datadog/profiling/component.rb

```diff
@@ -372,6 +375,35 @@ def self.build_profiler_component(settings:, agent_settings:, optional_tracer:)
           2.0
         end
       end
+
+      # To add just a bit more complexity to our detection code, in https://github.com/DataDog/dd-trace-rb/issues/3334
+      # a user reported that our code was incorrectly flagging the mariadb variant of libmysqlclient as being
+      # incompatible. In fact we have no reports of the mariadb variant needing the "no signals" workaround,
+      # so we flag it as compatible when it's in use.
+      #
+      # A problem is that there doesn't seem to be an obvious way to query the mysql2 gem on which kind of
+      # libmysqlclient it's using, so we detect it by looking at the version.
+      #
+      # The info method for mysql2 with mariadb looks something like this:
+      # `{:id=>30308, :version=>"3.3.8", :header_version=>"11.2.2"}`
+      #
+      # * The version seems to come from https://github.com/mariadb-corporation/mariadb-connector-c and the latest
+      # one is 3.x.
+      # * The header_version is what people usually see as the "mariadb version"
+      #
+      # As a comparison, for libmysql the info looks like:
+      # * `{:id=>80035, :version=>"8.0.35", :header_version=>"8.0.35"}`
+      #
+      # Thus our detection is version 4 or older, because libmysqlclient 4 is almost 20 years old so it's most probably
+      # not that one + header_version being 10 or newer, since according to https://endoflife.date/mariadb that's a
+      # sane range for modern mariadb releases.
+      private_class_method def self.looks_like_mariadb?(info, libmysqlclient_version)
+        header_version = Gem::Version.new(info[:header_version]) if info[:header_version]
+
+        !!(header_version &&
+          libmysqlclient_version < Gem::Version.new('5.0.0') &&
```

Comment:
Based on https://github.com/brianmario/mysql2/blob/79f78f940685396f1b5f30ec502544bb7e3ba9cf/ext/mysql2/client.c#L148 it seems like one may be able to use `mysql2` without linking/using the connector-c library in which case the reported version may be `>= 10.0.0` even for MariaDB.

Doesn't seem like they have any better way to distinguish between mariadb and mysql themselves in the gem based on the version gymnastics being done around that line.

So looks like eventually any logic we add here will get out of date. 

How crazy would it be to add a settings knob to complement this auto-detection and enabling users to force-set one flavour or the other?

Reply:
Ah, good call out!

Although looking at the code:

```c
  if (version >= 100000                         // MariaDB (all versions numbered 10.x)
    || (version >= 30000 && version < 40000)    // MariaDB Connector/C (all versions numbered 3.x)
```

...I think we'd be safe, although somewhat by accident.

The problem in #3334 is that the version check is testing for libmysqlclient older than 8.x and using that to choose to enable the "no signals" workaround; because mariadb connector/c reports 3.x, the logic would trigger.

But if for mariadb without connector/c the version reported is 10.x, we're fine there too, because it's >= 8.x.

> How crazy would it be to add a settings knob to complement this auto-detection and enabling users to force-set one flavour or the other?

There's already a [`settings.advanced.no_signals_workaround_enabled`](https://github.com/DataDog/dd-trace-rb/blob/7088ace6a50857b7e0931fa5eb67587ad0cae3d3/lib/datadog/core/configuration/settings.rb#L410) that can be set to `true` or `false` to override the autodetection.

Were you thinking of something like that or did you mean having something more specific to only override the mysql2 detection?

Reply:
Ah right. The workaround applies globally so no point making anything specific to mysql right? Just need to be aware of it when dealing with eventual support cases 👍

Reply:
Yeah the global one has worked so far. It's also documented in https://docs.datadoghq.com/profiler/profiler_troubleshooting/ruby/#unexpected-failures-or-errors-from-ruby-gems-that-use-native-extensions-in-dd-trace-rb-1110 .